lib/evmone/eof.cpp

```diff
@@ -90,37 +166,285 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
+
+    if (type_section_sizes.empty())
+        return std::vector{EOF1TypeHeader{0, 0, 0}};
+
+    std::vector<EOF1TypeHeader> types;
+
+    // guaranteed by EOF headers validation
+    assert(header_size + type_section_sizes[0] < container.size());
+
+    for (auto offset = header_size; offset < header_size + type_section_sizes[0]; offset += 4)
+    {
+        types.emplace_back(EOF1TypeHeader{container[offset], container[offset + 1],
+            static_cast<uint16_t>(container[offset + 2] << 8 | container[offset + 3])});
+    }
+
+    // check 1st section is (0, 0)
+    if (types[0].inputs_num != 0 || types[0].outputs_num != 0)
+        return EOFValidationError::invalid_first_section_type;
+
+    for (const auto& t : types)
+    {
+        if (t.max_stack_height > MAX_STACK_HEIGHT)
+            return EOFValidationError::max_stack_height_above_limit;
+
+        if (t.outputs_num > OUTPUTS_INPUTS_NUMBER_LIMIT ||
+            t.inputs_num > OUTPUTS_INPUTS_NUMBER_LIMIT)
+            return EOFValidationError::inputs_outputs_num_above_limit;
+    }
+
+    return types;
+}
+
 EOFValidationError validate_instructions(evmc_revision rev, bytes_view code) noexcept
 {
     assert(!code.empty());  // guaranteed by EOF headers validation
 
+    const auto& cost_table = baseline::get_baseline_cost_table(rev, 1);
+
     size_t i = 0;
-    uint8_t op = code[0];
+    uint8_t op = 0;
     while (i < code.size())
     {
         op = code[i];
-        const auto& since = instr::traits[op].since;
-        if (!since.has_value() || *since > rev)
+        if (cost_table[op] == instr::undefined)
             return EOFValidationError::undefined_instruction;
 
-        i += instr::traits[op].immediate_size;
+        if (op == OP_RJUMPV)
+        {
+            if (i + 1 < code.size())
+            {
+                const auto count = code[i + 1];
+                if (count < 1)
+                    return EOFValidationError::invalid_rjumpv_count;
+                i += static_cast<size_t>(1 /* count */ + count * 2 /* tbl */);
+            }
+            else
+                return EOFValidationError::truncated_instruction;
+        }
+        else
+            i += instr::traits[op].immediate_size;
+
+        if (i >= code.size())
+            return EOFValidationError::truncated_instruction;
+
         ++i;
     }
 
-    if (!instr::traits[op].is_terminating)
-        return EOFValidationError::missing_terminating_instruction;
-
     return EOFValidationError::success;
 }
 
+bool validate_rjump_destinations(
+    const EOF1Header& header, size_t code_idx, bytes_view::const_iterator container) noexcept
+{
+    // Collect relative jump destinations and immediate locations
+    std::vector<size_t> rjumpdests;
+    const auto code_size = header.code_sizes[code_idx];
+    std::vector<bool> immediate_map(code_size);
+    for (size_t i = 0; i < code_size; ++i)
+    {
+        const auto op_pos = header.code_begin(code_idx) + i;
```

Comment:
minor simplification: you could use `op_pos` as a loop variable and eliminate `i`

Reply:
or maybe better idea: define `bytes_view code{&container[header.code_begin(code_idx)], code_size}` before the loop and use it instead of calling `code_begin()` inside the loop.